File: src/main/java/de/steklopod/sec01/Lec06SupplierRefactoring.java
package de.steklopod.sec01;

import de.steklopod.courseutil.Util;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

public class Lec06SupplierRefactoring {

    public static void main(String[] args) {

        getName();
        String name = getName()
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        System.out.println(name);
        getName();

        Util.sleepSeconds(4);
    }

    private static Mono<String> getName(){
        System.out.println("entered getName method");
        return Mono.fromSupplier(() -> {
            System.out.println("Generating name..");
            Util.sleepSeconds(3);
            return Util.faker().name().fullName();
        }).map(String::toUpperCase);
    }


}

File: src/main/kotlin/de/steklopod/sec01/Lec06SupplierRefactoring.kt
package de.steklopod.sec01

import de.steklopod.utils.Util
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import java.util.*

object Lec06SupplierRefactoring {
    @JvmStatic
    fun main(args: Array<String>) {
        name

        val nameGenerated: String? = name
            .subscribeOn(Schedulers.boundedElastic())
            .block()
        println(nameGenerated)

        name
        Util.sleepSeconds(4)
    }

    private val name: Mono<String>
        get() {
            println("entered getName method")
            return Mono.fromSupplier {
                println("Generating name..")
                Util.sleepSeconds(3)
                Util.faker().name().fullName()
            }.map { obj: String -> obj.uppercase(Locale.getDefault()) }
        }
}
